### DISCUSSION TITLE
How to debug edge devices in FlowForge?

### DISCUSSION THREAD

**Discussion by Sulomus** [FlowFuse/flowfuse#1566] [Q&A] [answered]
Hello,
I am experimenting now with FlowForge and FlowForge’s device agent.
I wanted to ask what are common ways to debug the FlowForge agent on the edge device?

1.	Are there any tools to debug the edge devices from the cloud?
2.	Can I somehow debug the edge device when I can get access to it (via SSH for example)?

I know about FlowForge Tools Plugin – but was wondering if there are some other ways how to debug the Edge Devices.

**Comment by knolleary** [comment, marked as answer]
Hi @Sulomus 

This is an area that we need to improve our capabilities around, and it does depend on what sort of debug activities you want to do.

We do not enable the editor on devices, so you cannot access the Debug sidebar directly, nor make changes on the device itself. If you have ssh access to the device, you can access the Device Agent and Node-RED log - so configuring your Debug nodes to also write to the system console would make their output more accessible with the current agent.

What sort of capabilities would you be looking for in this area?

**Reply by Sulomus** [reply]
> If you have ssh access to the device, you can access the Device Agent and Node-RED log

The debug log can be one of the first things I can try. 

> What sort of capabilities would you be looking for in this area?

Our team is directed into the area of monitoring 3rd party devices (Most commonly we are dealing with PLC's in industry, but we have more and more IoT type customers where we need to deploy 10s and 100s of monitoring devices- thus we want to use FlowForge! 😄 ). 

The problem now I am facing is that when I deploy a flow I can't know if the communication nodes are correctly configured, if they connected to the 3rd party device, or if there is any sort of unexpected behavior (when for example playing with the local file system). 

I don't know if can make a "feature request" here but I'll try:

1) I think one of the ways how to debug a EdgeDevice would be to send it's log into the FlowForge platform. Now you can only see the FlowForge docker log
2) Another awesome feature would be to "switch" between debug consoles in the FlowForge editor - so you could see what is specific EdgeDevice outputing. 
- I think it would be great if there would be also an option of "grouping" in the debug console in the FlowForge editor.
- For example you deploy a flow and want to see what are the devices output. And in the debug console you would have info like: 90% of EdgeDevices sended this payload: "ABC ABC" and 10% of EdgeDevices sended this payload: "Error 1"

**Reply by knolleary** [reply]
@Sulomus thanks for the feedback. That's really useful.

I think your suggested features are aligned with the type of things we had in mind in this area, but I don't think we'd got as far to capture them in a specific issue. I'll get something raised and link back to it from here.
